### PR TITLE
Move the model creation to the last step before trainer creation

### PR DESCRIPTION
**Pull request by dakinggg**
The model download from huggingface is often quite slow, so this PR moves the model creation to the last step before the trainer creation. This way any other misconfigurations or dataset init validation will fail faster without having to wait for the model download from huggingface.

Before with bad data path: `fast-fail-data-before-1-5KeAKO` (fails after model download)
After with bad data path: `fast-fail-data-after-1-HMYrlQ` (fails before model download)
After with normal data path: `fast-fail-data-normal-3-5YV31Z` (trains and evals as normal)

Closes [GRT-2320](https://mosaicml.atlassian.net/browse/GRT-2320)

### CODE REVIEW

File: scripts/train/train.py [hunk 1]
```diff
@@ -442,22 +423,50 @@ def main(cfg: DictConfig):
...
+    optimizer = build_optimizer(model, optimizer_name, optimizer_config)
+
+    # Now add the eval metrics
+    if eval_loader_config is not None:
+        assert eval_loader is not None
+        assert model.train_metrics is not None
+        eval_metric_names = list(model.train_metrics.keys())
+        eval_loader.metric_names = eval_metric_names
+        evaluators = [eval_loader
+                     ] + evaluators  # Put the base eval_loader first
```
Comment (mvpatel2000) [inline scripts/train/train.py:468]:
```suggestion
        evaluators.insert(eval_loader, 0)  # Put the base eval_loader first
```
I think this is a bit easier to parse and fits on one line. Optional

**Review by mvpatel2000 — APPROVED** [review]
LGTM. Good idea

File: scripts/train/train.py [hunk 2]
```diff
@@ -442,22 +423,50 @@ def main(cfg: DictConfig):
...
+    n_params = sum(p.numel() for p in model.parameters())
+    logged_cfg.update({'n_params': n_params})
+
+    # Optimizer
+    optimizer_name: str = optimizer_config.pop('name')
+    optimizer = build_optimizer(model, optimizer_name, optimizer_config)
+
+    # Now add the eval metrics
+    if eval_loader_config is not None:
+        assert eval_loader is not None
```
Comment (j316chuck) [inline scripts/train/train.py:463]: nit: should we raise a ValueError here?

File: tests/test_train_inputs.py [hunk 3]
```diff
@@ -17,6 +18,38 @@
 from scripts.train.train import main  # noqa: E402
 
 
+def make_fake_index_file(path: str) -> None:
```
Comment (j316chuck) [inline tests/test_train_inputs.py:21]: Thanks making this mock data function!

**Review by j316chuck — APPROVED** [review]
lgtm ✅

File: scripts/train/train.py (hunk 2 above)
Comment (dakinggg) [reply scripts/train/train.py:463]: nah, this is just for pyright

**Pull request merged**